site/content/ytt/docs/develop/how-to-modularize.md

```diff
@@ -0,0 +1,481 @@
+    - app.kubernetes.io/name: #@ name
+spec:
+  selector:
+    matchLabels:
+      app: #@ name
+  replicas: #@ replicas
+  template:
+    metadata:
+      labels:
+        - app.kubernetes.io/version: #@ version
+        - app.kubernetes.io/name: #@ name
+    spec:
+      containers:
+        - name: #@ name
+          image: docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: #@ name + "-service"
+  labels:
+    - app.kubernetes.io/version: #@ version
+    - app.kubernetes.io/name: #@ name
+spec:
+  type: ClusterIP
+  ports:
+    - port: 38080
+      targetPort: 8080
+```
+Execute this template by running `ytt -f config.yml`.
+
+The result is identical to our original template, and now we can be sure all our repeated values will be consistent and easier to modify.
+
+### Functions
+[Functions](lang-ref-def.md) provide a way to extract common code into a separate fragment or code snippet.
+
+There are two ways to define a function in ytt, a Starlark function, or a yaml fragment function. We will use both here.
+
+Starlark functions make use of a `return` statement. Because of this they can be great for returning a value that must be transformed in some way.
+
+YAML fragment functions differ in that they are YAML structure wrapped in a Starlark function definition. Everything inside the function will be the return value. They can be great when needing to return nested YAML structure, or key and value pairs.
+
+Going back to the previous solution, we can see each `labels` key is duplicated YAML, like `app.kubernetes.io/version: #@ version`. There is also some duplicated string manipulation in the `metadata.name` key.
+```yaml
+#! config.yml
+
+#@ name = "frontend"
+#@ namespace = "default"
+#@ version = 0.1.0
+#@ replicas = 1
+
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: #@ name + "-deployment"
+  namespace: #@ namespace
+  labels:
+    - app.kubernetes.io/version: #@ version
+    - app.kubernetes.io/name: #@ name
+spec:
+  selector:
+    matchLabels:
+      app: #@ name
+  replicas: #@ replicas
+  template:
+    metadata:
+      labels:
+        - app.kubernetes.io/version: #@ version
+        - app.kubernetes.io/name: #@ name
+    spec:
+      containers:
+        - name: #@ name
+          image: docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: #@ name + "-service"
+  labels:
+    - app.kubernetes.io/version: #@ version
+    - app.kubernetes.io/name: #@ name
+spec:
+  type: ClusterIP
+  ports:
+    - port: 38080
+      targetPort: 8080
+```
+
+Move the duplicated `labels` keys into a YAML fragment function, and move name calculation into a Starlark function.
+
+```yaml
+#! config.yml
+
+#@ name = "frontend"
+#@ namespace = "default"
+#@ version = "0.1.0"
+#@ replicas = 1
+
+#! Starlark function
+
```

Comment:
nit: lets remove newline between comment and function. this will make it look a bit more "together"